lib/irb.rb

```diff
@@ -560,6 +595,19 @@ def eval_input
           begin
             # Assignment expression check should be done before evaluate_line to handle code like `a /2#/ if false; a = 1`
             is_assignment = assignment_expression?(line)
+
+            # If the integration with debugger is activated, we need to handle certain input differently
+            if @context.with_debugger
+              command_class = load_command_class(line)
+              # First, let's pass debugging command's input to debugger
+              # Secondly, we need to let debugger evaluate non-command input
+              # Otherwise, the expression will be evaluated in the debugger's main session thread
+              # This is the only way to run the user's program in the expected thread
+              if !command_class || command_class.ancestors.include?(ExtendCommand::DebugCommand)
+                return line
+              end
+            end
+
             evaluate_line(line, line_no)
```

Comment:
After evaluating line, the context may be changed to debug mode.

With this code
```ruby
class Foo
  def bar()
    binding.irb
  end
  def to_s() = 'foo'
end
binding.irb
```

```ruby
From: a.rb @ line 7 :

    2:   def bar()
    3:     binding.irb
    4:   end
    5:   def to_s() = 'foo'
    6: end
 => 7: binding.irb
irb(main):001:0> Foo.new.bar

From: a.rb @ line 3 :

    1: class Foo
    2:   def bar()
 => 3:     binding.irb
    4:   end
    5:   def to_s() = 'foo'
    6: end
    7: binding.irb
irb(foo):001:0> debug         # Enter debug mode.
irb:rdbg(foo):002:0> continue # Goes back to line 7. Context has changed to debug mode.
=> #<Class:DEBUGGER__>
irb(main):002:0>              # Still in debug mode, but prompt looks not a debug mode.
irb(main):003:0> whereami     # Location is wrong.

From: a.rb @ line 3 :

    1: class Foo
    2:   def bar()
 => 3:     binding.irb
    4:   end
    5:   def to_s() = 'foo'
    6: end
    7: binding.irb
=> nil
irb(main):004:0> Foo.new.bar

From: a.rb @ line 3 :

    1: class Foo
    2:   def bar()
 => 3:     binding.irb
    4:   end
    5:   def to_s() = 'foo'
    6: end
    7: binding.irb
irb:rdbg(foo):003:0>          # Reached to next binding.irb. Prompt shows that it's in debug mode.
```


I think we need something like this after `evaluate_line`.
```ruby
if !@context.with_debugger && defined?(DEBUGGER__::SESSION)
  debug setup and throw :IRB_EXIT
end
```

Reply:
Thanks for catching this case 👍 
I think the root problem is that IRB doesn't support nested sessions as described in https://github.com/ruby/irb/issues/650 so I plan to address it separately.